text/0031-corejam.md

```diff
@@ -0,0 +1,681 @@
+
+```rust
+fn is_authorized(param: &AuthParam, package: &WorkPackage, core_index: CoreIndex) -> bool;
+```
+
+This function is executed in a metered VM and subject to a modest system-wide limitation on execution time. If it overruns this limit or panicks on some input, it is considered equivalent to returning `false`. While it is mostly stateless (e.g. isolated from any Relay-chain state) it is provided with the package's `context` field in order to give information about a recent Relay-chain block. This allows it to be provided with a concise proof over some recent Relay-chain state.
+
+A single `Authorizer` value is associated with the index of the Core at a particular Relay-chain block and limits in some way what Work Packages may be legally processed by that Core.
+
+Since encoded `Authorizer` values may be fairly large (up to 1,038 bytes here), they may not be a drop-in replacement for the `ParaId`/`TaskId` used at present in the Agile Coretime interface. Because of this, we provide a lookup mechanism allowing a much shorter `AuthId` to be used within the Coretime scheduling messaging. Conveniently, this is precisely the same datatype size (32-bit) as a `ParaId`/`TaskId`.
+
+There is an Authorizations Pallet which stores the association. Adding a new piece of code is permissionless but requires a deposit commensurate with its size.
+
+```rust
+type AuthId = u32;
+type Authorizers = StorageMap<AuthId, Authorizer>;
+```
+
+An *Authorization* is simply a blob which helps the Authorizer recognize a properly authorized Work Package. No constraints are placed on Authorizers over how they may interpret this blob. Expected authorization content includes signatures, Merkle-proofs and more exotic succinct zero-knowledge proofs.
+
+_(Note: depending on future Relay-chain Coretime scheduling implementation concerns, a window of Relay-chain blocks)._
+
+The need of validators to be rewarded for doing work they might reasonably expect to be useful competes with that of the Coretime procurers to be certain to get work done which is useful to them. In Polkadot 1.0, validators only get rewarded for PoVs ("work packages") which do not panic or overrun. This ensures that validators are well-incentivized to ensure that their computation is useful for the assigned parachain. This incentive model works adequately where all PVF code is of high quality and collators are few and static.
+
+However with this proposal (and even the advent of on-demand parachains), validators have little ability to identify a high-quality Work Package builder and the permissionless design means a greater expectation of flawed code executing in-core. Because of this, we make a slightly modified approach: Work Packages must have a valid Authorization, i.e. the Coretime-assigned `is_authorized` returns `true` when provided with the Work Package. However, Validators get rewarded for *any* such authorized Work Package, even one which ultimately panics or overruns on its evaluation.
```

Comment:
In the document, it is mentioned that for any work package which overruns or panics, the `is_authorized` will return false.

```
If it overruns this limit or panics on some input, it is considered equivalent to returning false.
```

However, here it is mentioned that validators will be rewarded for any work package that is authorized, so isn't the following sentence contradictory?

```
However, Validators get rewarded for *any* such authorized Work Package, even one which ultimately panics or overruns on its evaluation.
```

I would assume that a work package cannot be authorized if it panics or overruns. Or I may not correctly understand what is meant by evaluation here.

Reply:
Yeah, authorization cannot know if some panic occurs later.  It's possible dot holders wind up rewarded for failed work.  We rewards precisely during disputes too.  We cannot reward backers for failed work, only successful work on a finalized fork of the relay chain.  We only approximately reward approval checkers for successful work on a finalized fork of the relay chain.

Reply:
> My intention with my comment wasn't to further discuss rewards here. I just noticed two statements in the RFC that seemed contradictory to me. In one place, it is mentioned that a work package that overruns or panics cannot be authorized, and then later it is mentioned that validators will be rewarded for any authorized work packages even if they overrun or panic.

An authorized Work Package which overruns is still reported. If the authorization code itself (which is not part of the WP) overruns or panics, then it is not authorized. The WP itself never executes.

Reply:
We cannot pay backers until after inclusion/integration of candidates, including execution.  We risk it being profitable to stall cores otherwise.

We cannot execute the overrunning work package either because this breaks other invariants, including some essential for wasm.

As noted above, backers represent only a relatively minor duty worth only like 2% of rewards.  We should not break other compoinents to reward backers more precisely.  Instead we should solve builder/collator spam via solfter means, like reputaiton or just expect projects solve it, given it'll usually be their buggy code.